thriftserver/server/src/main/scala/org/apache/livy/thriftserver/LivyThriftSessionManager.scala

```
@@ -0,0 +1,609 @@
+  }
+
+  private def incrementManagedSessionActiveUsers(livySessionId: Int): Unit = synchronized {
+    managedLivySessionActiveUsers(livySessionId) = numberOfActiveUsers(livySessionId) + 1
+  }
+
+  def onUserSessionClosed(sessionHandle: SessionHandle, livySession: InteractiveSession): Unit = {
+    val closeSession = synchronized[Boolean] {
+      val activeUsers = managedLivySessionActiveUsers(livySession.id)
+      if (activeUsers == 1) {
+        // it was the last user, so we can close the LivySession
+        managedLivySessionActiveUsers -= livySession.id
+        true
+      } else {
+        managedLivySessionActiveUsers(livySession.id) = activeUsers - 1
+        false
+      }
+    }
+    if (closeSession) {
+      server.livySessionManager.delete(livySession)
+    } else {
+      // We unregister the session only if we don't close it, as it is unnecessary in that case
+      val rpcClient = new RpcClient(livySession)
+      try {
+        rpcClient.executeUnregisterSession(sessionHandle).get()
+      } catch {
+        case e: Exception => warn(s"Unable to unregister session $sessionHandle", e)
+      }
+    }
+  }
+
+  /**
+   * If the user specified an existing sessionId to use, the corresponding session is returned,
+   * otherwise a new session is created and returned.
+   */
+  private def getOrCreateLivySession(
+      sessionHandle: SessionHandle,
+      sessionId: Option[Int],
+      username: String,
+      createLivySession: () => InteractiveSession): InteractiveSession = {
+    sessionId match {
+      case Some(id) =>
+        server.livySessionManager.get(id) match {
+          case None =>
+            warn(s"Session id $id doesn't exist, so we will ignore it.")
+            createLivySession()
+          case Some(session) if !server.isAllowedToUse(username, session) =>
+            warn(s"Session id $id doesn't belong to $username, so we will ignore it.")
+            createLivySession()
```

Seems like this should result in an error, not in a new session being created?

---

The some problem would happen if the user specifies a non-existing session id. Then we should throw an error in that case too. And the same when the session is not active anymore. What do you think?

---

+1 for erroring out in all cases where the user expected session is not available.